keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,222 @@
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Infrastructure Needed](#infrastructure-needed)
+
+
+## Summary
+
+We propose to use operators for managing cluster addons.  Each addon will have
+its own CRD, and users will be able to perform limited tailoring of the addon
+(install/don’t install, choose version, primary feature selection) by modifying
+the CR.  The operator encodes any special logic (e.g. dependencies) needed to
+install the addon.
+
+By creating a CRD per addon, we are able make use of the kubernetes API
+machinery for those per-addon options.
+
+We will create tooling to make it easy to build addon operators that follow the
+best practices we identify as part of this work.  For example, we expect that
+most addons will be declarative, and likely be specified as part of a "cluster
+bundle", so we will make it easy to build basic addon operators that follow
+these patterns.
+
+We hope that components will choose to maintain their own operators, encoding
+their knowledge of how best to operate their addon.
+
+
+## Motivation
+
+Addons are components that are managed alongside the lifecycle of the cluster.
+They are often tied to or dependent on the configuration of other cluster
+components.  Management of these components has proved complicated.  Our
+existing solution in the form of the bash addon-manager has many known
+shortcomings and is not widely adopted.  As we focus more development outside of
+the kubernetes/kubernetes repo, we expect more addon components of greater
+complexity.  This is one of the long-standing backlog items for
+sig-cluster-lifecycle.
+
+Use of operators is now generally accepted, and the benefits to other
+applications are generally recognized.  We aim to bring the benefits of
+operators to addons also.
+
+### Goals
+
+* Explore the use of operators for managing addons
+* Create patterns, libraries & tooling so that addons are of high quality,
+  consistent in their API surface (common fields on CRDs, use of Application
+  CRD, consistent labeling of created resources), yet are easy to build.
+* Build addons for the basic set of components, acting as a quality reference
+  implementation suitable for production use.  We aim also to demonstrate the
+  utility and explore any challenges, and to verify that the tooling does make
+  addon-development easy.
+
+
+### Non-Goals
+
+* We do not intend to mandate that all installation tools use addon operators;
+  installation tools are free to choose their own path.
+* Management of non-addons is out of scope (for example installation of end-user
+  applications, or of packaged software that is not an addon)
+
+
+## Proposal
+
+This is the current plan of action; it is based on experience gathered and work
+done for Google’s GKE-on-prem product.  However we don’t expect this will
```

Comment:
openshift has been exploring this space as well.

Reply:
Is there a summary of your learnings written down?